pkgs/dart_mcp/lib/src/api/tools.dart

```diff
@@ -528,5 +1441,158 @@ extension type ListSchema.fromMap(Map<String, Object?> _value)
+        if (seenItems.contains(item)) {
+          duplicates.add(item);
+        } else {
+          seenItems.add(item);
+        }
+      }
+      isValid = false;
+      accumulatedFailures.add(
+        ValidationError(
+          ValidationErrorType.uniqueItemsViolated,
+          path: currentPath,
+          details: 'List contains duplicate items: ${duplicates.join(', ')}.',
+        ),
+      );
+    }
+
+    final evaluatedItems = List<bool>.filled(data.length, false);
+    if (prefixItems case final pItems?) {
+      for (var i = 0; i < pItems.length && i < data.length; i++) {
+        evaluatedItems[i] = true;
+        final itemPath = [...currentPath, i.toString()];
+        final prefixItemSpecificFailures = _createHashSet();
+        if (!pItems[i]._validateSchema(
+          data[i],
+          itemPath,
+          prefixItemSpecificFailures,
+        )) {
+          isValid = false;
+          accumulatedFailures.add(
+            ValidationError(
+              ValidationErrorType.prefixItemInvalid,
+              path: itemPath,
+            ),
+          );
+          accumulatedFailures.addAll(prefixItemSpecificFailures);
+        }
+      }
+    }
+    if (items case final itemSchema?) {
+      final startIndex = prefixItems?.length ?? 0;
+      for (var i = startIndex; i < data.length; i++) {
+        evaluatedItems[i] = true;
+        final itemPath = [...currentPath, i.toString()];
+        final itemSpecificFailures = _createHashSet();
+        if (!itemSchema._validateSchema(
+          data[i],
+          itemPath,
+          itemSpecificFailures,
+        )) {
+          isValid = false;
+          accumulatedFailures.add(
+            ValidationError(ValidationErrorType.itemInvalid, path: itemPath),
+          );
+          accumulatedFailures.addAll(itemSpecificFailures);
+        }
+      }
+    }
+    if (unevaluatedItems == false) {
+      for (var i = 0; i < data.length; i++) {
+        if (!evaluatedItems[i]) {
+          final itemPath = [...currentPath, i.toString()];
+          isValid = false;
+          accumulatedFailures.add(
+            ValidationError(
+              ValidationErrorType.unevaluatedItemNotAllowed,
+              path: itemPath,
+              details: 'Unevaluated item in list at index $i',
+            ),
+          );
+          // Only report the first unevaluated item to avoid excessive errors.
+          // If we want all, remove the break. For now, keeping existing
+          // behavior of early exit.
+          break;
+        }
+      }
+    }
+    return isValid;
+  }
+}
+
+HashSet<ValidationError> _createHashSet() {
+  return HashSet<ValidationError>(
+    equals: (ValidationError a, ValidationError b) {
+      return const ListEquality<String>().equals(a.path, b.path) &&
+          a.details == b.details &&
+          a.error == b.error;
+    },
+    hashCode: (ValidationError error) {
+      return Object.hashAll([...error.path ?? [], error.details, error.error]);
```

Comment:
```suggestion
      return Object.hashAll([...error.path ?? const [], error.details, error.error]);
```

Reply:
I would also expect the total amount of errors here to be low. That won't always be the case though, but I am more concerned about the performance of checking for the errors, than the actual handling of the errors, and in theory these hash functions should not get called a bunch of times on any given instance.